shared/p2p/service.go

```diff
@@ -91,6 +77,85 @@ func (s *Server) Stop() error {
 	return nil
 }
 
+// RegisterTopic with a message and the adapter stack for the given topic. The
+// message type provided will be feed selector for emitting messages received
+// on a given topic.
+//
+// The topics can originate from multiple sources. In other words, messages on
+// TopicA may come from direct peer communication or a pub/sub channel.
+func (s *Server) RegisterTopic(topic string, message interface{}, adapters ...Adapter) {
+	msgType := reflect.TypeOf(message)
+	log.WithFields(logrus.Fields{
+		"topic": topic,
+	}).Debug("Subscribing to topic")
+
+	s.topicMapping[msgType] = topic
+
+	sub, err := s.gsub.Subscribe(topic)
+	if err != nil {
+		log.Errorf("Failed to subscribe to topic: %v", err)
+		return
+	}
+	feed := s.Feed(msgType)
+
+	// Reverse adapter order
+	for i := len(adapters)/2 - 1; i >= 0; i-- {
+		opp := len(adapters) - 1 - i
+		adapters[i], adapters[opp] = adapters[opp], adapters[i]
+	}
+
+	go (func() {
+		defer sub.Cancel()
+		for {
+			msg, err := sub.Next(s.ctx)
+
+			if s.ctx.Err() != nil {
+				log.WithError(s.ctx.Err()).Debug("Context error")
+				return
+			}
+
+			if err != nil {
+				log.Errorf("Failed to get next message: %v", err)
+				return
+			}
+
+			var h Handler = func(ctx context.Context, pMsg Message) {
+				s.emit(feed, msg, msgType)
+			}
+
+			pMsg := Message{}
+
+			for _, adapter := range adapters {
+				h = adapter(h)
+			}
+
+			ctx, _ := context.WithTimeout(s.ctx, 10*time.Second)
```

Comment:
Why? There is a timeout, i'm not sure we even need this timeout. I'll just remove it.

Reply:
https://stackoverflow.com/questions/44393995/what-happens-if-i-dont-cancel-a-context

Looks like the timeout exists so that the adapter doesn't run continuously for more than 10 seconds.